### PR TITLE
Add boundary exception handling

### PR DESCRIPTION
**Pull request by henry2004y**
Trying to handle #564, with some other minor additional changes:
1. Add exception handling for unknown boundary types during initialization;
2. Remove Boolean type comparison with equal sign;
3. Formatting adjustment 

With the new PR, unrecognized boundary type names listed under command `[boundaries]` in the configuration file will trigger `MPI_abort` during initialization:
```shell
Filtering is off and max number of Passes is = 	 0
Parameter data file (sw1.dat) has 1 values
Parameter data file (sw1.dat) has 1 values
Unknown type of boundary read: haha
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 1 in communicator MPI_COMM_WORLD
with errorcode 1.
```

### CODE REVIEW

File: sysboundary/sysboundary.cpp [hunk 1]
```diff
@@ -146,151 +144,192 @@ bool SysBoundary::addSysBoundary(
...
+                    "option boundaries.boundary, are you sure this is correct?"
+                 << endl;
       }
-      if(!isPeriodic[2] && !Readparameters::helpRequested) {
-         if(myRank == MASTER_RANK) cerr << "You set boundaries.periodic_z = no but you didn't load any system boundary condition using the option boundaries.boundary, are you sure this is correct?" << endl;
+      if (!isPeriodic[2] && !Readparameters::helpRequested) {
+         if (myRank == MASTER_RANK)
+            cerr << "You set boundaries.periodic_z = no but you didn't load any system boundary condition using the "
+                    "option boundaries.boundary, are you sure this is correct?"
+                 << endl;
```
Comment (ykempf) [inline sysboundary/sysboundary.cpp:181]: And also in z.

File: sysboundary/sysboundary.cpp [hunk 2]
```diff
@@ -146,151 +144,192 @@ bool SysBoundary::addSysBoundary(
...
+               cerr << "You load Outflow system boundary conditions on the y+ or y- face but there is not enough cells "
+                       "in that direction to make sense."
+                    << endl;
             exit(1);
          }
-         if((faces[4] || faces[5]) && P::zcells_ini < 5) {
-            if(myRank == MASTER_RANK) cerr << "You load Outflow system boundary conditions on the z+ or z- face but there is not enough cells in that direction to make sense." << endl;
+         if ((faces[4] || faces[5]) && P::zcells_ini < 5) {
+            if (myRank == MASTER_RANK)
+               cerr << "You load Outflow system boundary conditions on the z+ or z- face but there is not enough cells "
```
Comment (ykempf) [inline sysboundary/sysboundary.cpp:229]: And all of those too. { }

File: sysboundary/sysboundary.cpp [hunk 3]
```diff
@@ -146,151 +144,192 @@ bool SysBoundary::addSysBoundary(
...
+         if ((faces[4] || faces[5]) && P::zcells_ini < 5) {
+            if (myRank == MASTER_RANK)
+               cerr << "You load Maxwellian system boundary conditions on the z+ or z- face but there is not enough "
+                       "cells in that direction to make sense."
+                    << endl;
             exit(1);
          }
+      } else {
+         if (myRank == MASTER_RANK)
+            cerr << "Unknown type of boundary read: " << *it << endl;
```
Comment (ykempf) [inline sysboundary/sysboundary.cpp:296]: Good, this is the essential change. And I gave up flagging all { } missing for if blocks above.

File: sysboundary/sysboundary.cpp [hunk 4]
```diff
@@ -312,91 +351,93 @@ bool SysBoundary::checkRefinement(dccrg::Dccrg<spatial_cell::SpatialCell,dccrg::
...
             // among its neighbors.
-            if( ( ix == 0 && iy == 0 && iz == 0 ) || !technicalGrid.get(x+ix,y+iy,z+iz)) {
+            if ((ix == 0 && iy == 0 && iz == 0) || !technicalGrid.get(x + ix, y + iy, z + iz)) {
                continue;
             }
-            
-            if(layer == 1 && technicalGrid.get(x+ix,y+iy,z+iz)->sysBoundaryFlag == sysboundarytype::NOT_SYSBOUNDARY) {
+
+            if (layer == 1 &&
+                technicalGrid.get(x + ix, y + iy, z + iz)->sysBoundaryFlag == sysboundarytype::NOT_SYSBOUNDARY) {
```
Comment (ykempf) [inline sysboundary/sysboundary.cpp:394]: I don't see why this is split over 2 lines if line 389 above is not.

File: sysboundary/sysboundary.cpp [hunk 5]
```diff
@@ -475,145 +513,143 @@ bool SysBoundary::classifyCells(dccrg::Dccrg<spatial_cell::SpatialCell,dccrg::Ca
...
+            technicalGrid.get(x, y, z)->SOLVE = 0;
+
+            array<int32_t, 3> globalIndices = technicalGrid.getGlobalIndices(x, y, z);
+
+            if (((globalIndices[0] == 0 || globalIndices[0] == fsGridDimensions[0] - 1) &&
+                 !this->isBoundaryPeriodic(0)) ||
+                ((globalIndices[1] == 0 || globalIndices[1] == fsGridDimensions[1] - 1) &&
+                 !this->isBoundaryPeriodic(1)) ||
+                ((globalIndices[2] == 0 || globalIndices[2] == fsGridDimensions[2] - 1) &&
+                 !this->isBoundaryPeriodic(2))) {
```
Comment (ykempf) [inline sysboundary/sysboundary.cpp:585]: This new formatting is less clear to read imho.

File: sysboundary/sysboundary.cpp [hunk 6]
```diff
@@ -625,78 +661,80 @@ bool SysBoundary::applyInitialState(
...
       phiprof::start(timer);
 
       // Compute Vlasov boundary condition on system boundary/process inner cells
       vector<CellID> localCells;
-      getBoundaryCellList(mpiGrid,mpiGrid.get_local_cells_not_on_process_boundary(SYSBOUNDARIES_EXTENDED_NEIGHBORHOOD_ID),localCells);
-   
-      #pragma omp parallel for
-      for (uint i=0; i<localCells.size(); i++) {
+      getBoundaryCellList(
+          mpiGrid, mpiGrid.get_local_cells_not_on_process_boundary(SYSBOUNDARIES_EXTENDED_NEIGHBORHOOD_ID), localCells);
```
Comment (ykempf) [inline sysboundary/sysboundary.cpp:707]: Why this format?

File: sysboundary/sysboundary.cpp [hunk 7]
```diff
@@ -625,78 +661,80 @@ bool SysBoundary::applyInitialState(
...
-   // First the small stuff without overlapping in an extended neighbourhood:
-   #warning TODO This now communicates in the wider neighbourhood for both layers, could be reduced to smaller neighbourhood for layer 1, larger neighbourhood for layer 2.
+/*Transfer along boundaries*/
+// First the small stuff without overlapping in an extended neighbourhood:
+#warning TODO This now communicates in the wider neighbourhood for both layers, could be reduced to smaller neighbourhood for layer 1, larger neighbourhood for layer 2.
    SpatialCell::set_mpi_transfer_type(
-      Transfer::CELL_PARAMETERS|
-      Transfer::POP_METADATA|
-      Transfer::CELL_SYSBOUNDARYFLAG,true);
+       Transfer::CELL_PARAMETERS | Transfer::POP_METADATA | Transfer::CELL_SYSBOUNDARYFLAG, true);
```
Comment (ykempf) [inline sysboundary/sysboundary.cpp:685]: Either all indented or all in one line, but just all arguments on another line is even weirder than what was before I think.

File: sysboundary/sysboundary.cpp [hunk 8]
```diff
@@ -625,78 +661,80 @@ bool SysBoundary::applyInitialState(
...
-      #pragma omp parallel for
-      for (uint i=0; i<boundaryCells.size(); i++) {
+      getBoundaryCellList(mpiGrid, mpiGrid.get_local_cells_on_process_boundary(SYSBOUNDARIES_EXTENDED_NEIGHBORHOOD_ID),
+                          boundaryCells);
+#pragma omp parallel for
+      for (uint i = 0; i < boundaryCells.size(); i++) {
          cuint sysBoundaryType = mpiGrid[boundaryCells[i]]->sysBoundaryFlag;
-         this->getSysBoundary(sysBoundaryType)->vlasovBoundaryCondition(mpiGrid, boundaryCells[i],popID,calculate_V_moments);
+         this->getSysBoundary(sysBoundaryType)
+             ->vlasovBoundaryCondition(mpiGrid, boundaryCells[i], popID, calculate_V_moments);
```
Comment (ykempf) [inline sysboundary/sysboundary.cpp:737]: I find these function-call linebreaks confusing.

**Review by ykempf — COMMENTED** [review]
The actual change I approve of, most white-space cleanup is OK too, but some of the white-space changes are not making it easier in my opinion...

File: sysboundary/sysboundary.cpp (hunk 4 above)
Comment (henry2004y) [reply sysboundary/sysboundary.cpp:394]: The splitting/non-spliiting is determined by max line width, which is set to 200.

**Comment by ursg** [comment]
Right. I guess the actual code-semantic content of this pull request is clearly approved by all sides.

Instead of letting it hang for longer due to formatting discussions, what I'm going to do now is:
- Revert the formatting changes that I find decrease legibility (I mostly agree with what Yann has written about that) on a local branch
- Merge that into dev (which Github will recognize as this pull request being merged, thus closing it).

We can then discuss the formatting details further, if we really care.

As a general remark: the code formatting guidelines are great, but use your own intuition about what makes the code more and what makes it less readable.

And if in doubt, use ```git add --patch``` to only commit parts of a changed file, if the majority of changes are semantically irrelevant to a pull request.

**Pull request merged**